### ISSUE TITLE
Consider enabling other indices than usize

### ISSUE THREAD

**Issue by workingjubilee** [rust-lang/portable-simd#166]
This would minimize contortions for casting between various index sizes.

See: https://github.com/rust-lang/rust/issues/89193#issuecomment-92627025
> Since there is a bit of chatter here on the use cases for performing a scatter or gather for indexes, I can give a little bit of context, since I do think it is somewhat informative.
> 
> In this case I am working on a columnar store. You can kind of think of the use case as a secondary index over columnar data. For just a single primary index, you have a list of row_ids Simd<usize, LANES> and you can perform a simd gather instruction to get all of the values at those indexes. However, if you have a secondary index the simd gather from that secondary index needs to produce a Simd<usize, LANES>, so that can be passed directly into the second of the gather functions over the primary keys. In this way you go from needed 2*LANES loads to perform a secondary index lookup over data to only needing 2 simd operations.
> 
> Alternatively it would be nice if the Simd libraries were updated to perform gathers using indexes as Simd<u32, LANES> or Simd<u64, LANES>. That would dodge this issue altogether and it would mean less copies in our columnar store. Many times columnar stores you keep offsets as u32 (or even smaller) to keep things compressed. And it is nice to be able to act directly on those indices rather than performing intermediate conversions to usize.

**Comment by calebzulawski** [comment]
Since in LLVM it's just using pointers, anyway, we can probably allow it to take any unsigned integer and the optimizer will handle it from there.

**Comment by dyv** [comment]
This would be awesome. I just ran into another case where I was operating on dictionary encoded values (flexibly encoded as u8 or u16), and then needed to convert into usize to perform lookups into the dictionary.

**Comment by calebzulawski** [comment]
It would probably be reasonable to provide a wrapping int conversion function

**Comment by workingjubilee** [comment]
In practice, the underlying API uses u32. Hrm...

**Comment by workingjubilee** [comment]
The choice of `usize` is actually arbitrary here. There is no immediately obvious reason to not use u32 as the upper bound and accept `Into<u32>`. The design is intended to allow us to make maximal usage of the actual machines, so that is the reason why: we get to choose between things like this in our design.

**Comment by calebzulawski** [comment]
I disagree that it's arbitrary--`usize` is the correct type for indexing arrays, which is effectively all `gather` is.  I personally don't think we should overcomplicate the interface, and shouldn't use any tricks with `Into`.  Once we have casts analogous to `as` (#116) it should be pretty painless to use any integer type.

**Comment by scottmcm** [comment]
`usize` is the type for indexing arrays (unfortunately; I wish we had a smaller index type), but `u32` is the type for ["indexing" bits](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_shl).  I think I could justifiably argue for either in SIMD -- thinking of lanes in simd ops like bits in bitwise ops also seems reasonable to me.

(Though really we'll probably keep things `usize` as the primary option because otherwise `as_array` and friends look like they'd be a huge mess.)
